File: occasum-core/src/main/java/pw/stamina/occasum/properties/enums/AbstractEnumParsingService.java
package pw.stamina.occasum.properties.enums;

import java.util.Objects;

public abstract class AbstractEnumParsingService<T extends Enum<T>> implements EnumParsingService<T> {
    protected final Class<T> enumClass;
    private T[] enumConstants;

    protected AbstractEnumParsingService(Class<T> enumClass) {
        Objects.requireNonNull(enumClass, "enumClass");
        this.enumClass = enumClass;
    }

    @Override
    public String toString(T value) {
        return value.name();
    }

    @Override
    public T[] getEnumConstants() {
        if (enumConstants == null) {
            enumConstants = enumClass.getEnumConstants();
        }

        return enumConstants;
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/properties/enums/AbstractEnumParsingService.kt
package pw.stamina.occasum.properties.enums

abstract class AbstractEnumParsingService<T : Enum<T>> protected constructor(protected val enumClass: Class<T>) : EnumParsingService<T> {

    override val enumConstants: Array<out T>
        get() = enumClass.enumConstants

    override fun toString(value: T): String {
        return value.name
    }
}
